File: src/utils/evCalculator.ts
import { americanToImplied, americanToDecimal } from './odds'

// ─── EV calculation ───────────────────────────────────────────────────────────
// All results are labeled "estimated" — not guaranteed projections.

export interface EVInputs {
  /** American odds (e.g. -110, +130) */
  odds: number
  /** Estimated true probability of winning, expressed as 0-100 */
  projected_probability: number
  /** Stake in units */
  stake?: number
}

export interface EVResult {
  implied_probability: number   // from odds
  break_even: number            // same as implied_probability
  projected_probability: number // model input
  probability_edge: number      // projected - implied (percentage points)
  decimal_odds: number
  profit_if_win: number         // per unit
  estimated_ev: number          // fractional: positive = +EV
  estimated_ev_pct: number      // as percentage
  expected_profit_units: number // per unit staked
}

/**
 * Calculate expected value given estimated probability and American odds.
 *
 * EV = (p_win × profit) - (p_lose × stake)
 *
 * All probability inputs/outputs are 0–100 range internally;
 * fractional values used in formulas.
 */
export function calculateEV(inputs: EVInputs): EVResult {
  const { odds, projected_probability, stake = 1 } = inputs

  const implied_probability = americanToImplied(odds)
  const decimal_odds = americanToDecimal(odds)
  const profit_if_win = (decimal_odds - 1) * stake
  const loss_if_lose = stake

  const p_win = projected_probability / 100
  const p_lose = 1 - p_win

  const ev = p_win * profit_if_win - p_lose * loss_if_lose
  const ev_pct = ev / stake

  return {
    implied_probability,
    break_even: implied_probability,
    projected_probability,
    probability_edge: projected_probability - implied_probability,
    decimal_odds,
    profit_if_win,
    estimated_ev: ev,
    estimated_ev_pct: ev_pct,
    expected_profit_units: ev / stake,
  }
}

/**
 * Estimate closing line value (CLV).
 * Positive = we got a better line than closing; negative = line moved against us.
 */
export function calculateCLV(
  original_odds: number,
  closing_odds: number,
  bet_type: 'over' | 'under' | 'spread' | 'moneyline' = 'moneyline',
): number {
  const original_impl = americanToImplied(original_odds)
  const closing_impl = americanToImplied(closing_odds)
  // Positive CLV: we paid less juice than closing (line moved in our direction)
  if (bet_type === 'under') {
    return closing_impl - original_impl
  }
  return original_impl - closing_impl
}

/**
 * Derive a simple estimated projected probability from projection edge.
 * This is a heuristic for the mock data layer — real implementation would
 * use a calibrated model.
 *
 * @param base_implied  - implied probability from odds (0-100)
 * @param edge_pct      - estimated probability edge in percentage points
 */
export function estimateProjectedProbability(
  base_implied: number,
  edge_pct: number,
): number {
  return Math.min(Math.max(base_implied + edge_pct, 0), 95)
}

File: backend/app/services/ev_calculator.py
"""
EV (Expected Value) calculation utilities.

All outputs are labeled "estimated" — not guaranteed projections.
Use for analysis only.
"""
from dataclasses import dataclass
from typing import Optional


def american_to_implied(odds: int) -> float:
    """Convert American odds to implied probability (0–100)."""
    if odds < 0:
        return abs(odds) / (abs(odds) + 100) * 100
    return 100 / (odds + 100) * 100


def american_to_decimal(odds: int) -> float:
    """Convert American odds to decimal odds."""
    if odds < 0:
        return 1 + 100 / abs(odds)
    return 1 + odds / 100


def decimal_to_american(decimal: float) -> int:
    """Convert decimal odds to American odds."""
    if decimal >= 2:
        return int((decimal - 1) * 100)
    return int(-100 / (decimal - 1))


def break_even_probability(odds: int) -> float:
    """Minimum win rate needed to profit at these odds."""
    return american_to_implied(odds)


@dataclass
class EVResult:
    implied_probability: float
    projected_probability: float
    probability_edge: float
    decimal_odds: float
    profit_if_win: float
    estimated_ev: float
    estimated_ev_pct: float


def calculate_ev(odds: int, projected_probability: float, stake: float = 1.0) -> EVResult:
    """
    Calculate expected value from odds and estimated probability.

    EV = (p_win × profit_if_win) - (p_lose × stake)

    Args:
        odds: American odds (e.g. -110, +130)
        projected_probability: Model probability estimate (0-100)
        stake: Bet size in units (default 1.0)

    Returns:
        EVResult with all calculation components
    """
    implied = american_to_implied(odds)
    decimal = american_to_decimal(odds)
    profit = (decimal - 1) * stake
    p_win = projected_probability / 100
    p_lose = 1 - p_win
    ev = p_win * profit - p_lose * stake

    return EVResult(
        implied_probability=implied,
        projected_probability=projected_probability,
        probability_edge=projected_probability - implied,
        decimal_odds=decimal,
        profit_if_win=profit,
        estimated_ev=ev,
        estimated_ev_pct=ev / stake if stake else 0,
    )


def calculate_clv(original_odds: int, closing_odds: int, is_under: bool = False) -> float:
    """
    Closing line value in probability points.
    Positive = we got a better price than closing.
    """
    orig_impl = american_to_implied(original_odds)
    close_impl = american_to_implied(closing_odds)
    if is_under:
        return close_impl - orig_impl
    return orig_impl - close_impl


def estimate_projected_probability(base_implied: float, edge_pct: float) -> float:
    """Heuristic: add edge to implied probability, capped at 95%."""
    return min(max(base_implied + edge_pct, 0), 95)
